sw/host/opentitanlib/src/test_utils/bitbanging/pwm.rs

```diff
@@ -0,0 +1,81 @@
+// Copyright lowRISC contributors (OpenTitan project).
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use super::Bit;
+use anyhow::Result;
+
+#[derive(Clone, Debug)]
+pub struct PwmPeriod {
+    pub period: std::time::Duration,
+    pub duty_cycle: f32,
+}
+
+pub mod decoder {
+    use super::*;
+
+    #[derive(Clone, Debug)]
+    struct Sample<const PIN: u8> {
+        raw: u8,
+    }
+
+    impl<const PIN: u8> Sample<PIN> {
+        fn pin(&self) -> Bit {
+            ((self.raw >> PIN) & 0x01).into()
+        }
+    }
+
+    pub struct Decoder<const PIN: u8> {
+        pub active_level: Bit,
+        pub sampling_period: std::time::Duration,
+    }
+
+    impl<const PIN: u8> Decoder<PIN> {
+        /// Loop sampling the pin until a rising edge is detected.
+        fn find_first_active_level<I>(&self, samples: &mut I) -> Result<()>
+        where
+            I: Iterator<Item = Sample<PIN>>,
+        {
+            let _ = samples
+                .by_ref()
+                .find(|sample| sample.pin() == Bit::Low)
```

Comment:
Should this be `!self.active_level` instead of hard-coded low?

Reply:
Actually, can you just call `decode_period` once initially and discard the result?